gcsio/src/main/java/com/google/cloud/hadoop/gcsio/DeleteFolderOperation.java

```diff
@@ -0,0 +1,237 @@
+package com.google.cloud.hadoop.gcsio;
+
+import static com.google.cloud.hadoop.gcsio.GoogleCloudStorageExceptions.createJsonResponseException;
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.api.client.googleapis.batch.json.JsonBatchCallback;
+import com.google.api.client.googleapis.json.GoogleJsonError;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.client.http.HttpHeaders;
+import com.google.cloud.hadoop.util.ApiErrorExtractor;
+import com.google.common.base.Strings;
+import com.google.common.flogger.GoogleLogger;
+import com.google.storage.control.v2.DeleteFolderRequest;
+import com.google.storage.control.v2.StorageControlClient;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentHashMap.KeySetView;
+import java.util.concurrent.LinkedBlockingQueue;
+import javax.annotation.Nonnull;
+
+public class DeleteFolderOperation {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  // Maximum number of times to retry deletes in the case of precondition failures.
+  private static final int MAXIMUM_PRECONDITION_FAILURES_IN_DELETE = 4;
+  private ApiErrorExtractor errorExtractor = ApiErrorExtractor.INSTANCE;
+  private GoogleCloudStorageOptions storageOptions;
+  private final KeySetView<IOException, Boolean> innerExceptions;
+  private List<FolderInfo> folders;
+  private BatchExecutor batchExecutor;
+  private StorageControlClient storageControlClient;
+  private BlockingQueue<FolderInfo> folderDeleteBlockingQueue;
+  private ConcurrentHashMap<String, Long> countOfChildren;
+
+  DeleteFolderOperation(
+      List<FolderInfo> folders,
+      GoogleCloudStorageOptions storageOptions,
+      StorageControlClient storageControlClient) {
+    this.folders = folders;
+    this.storageOptions = storageOptions;
+    this.storageControlClient = storageControlClient;
+    this.folderDeleteBlockingQueue = new LinkedBlockingQueue<>(folders.size());
+
+    // threads for parallel delete of folder resources
+    this.batchExecutor = new BatchExecutor(this.storageOptions.getBatchThreads());
+
+    // Gather exceptions to wrap in a composite exception at the end.
+    this.innerExceptions = ConcurrentHashMap.newKeySet();
+
+    // Map to store number of children for each parent object
+    this.countOfChildren = new ConcurrentHashMap<>();
+  }
+
+  /** Helper function that performs the deletion process for folder resources */
+  public void performDeleteOperation() throws IOException {
```

Comment:
Looks like we are not throwing any excpetion if the async deletion fails?

Reply:
All exceptions are thrown by impl class.